### ISSUE TITLE
[Question] Ensuring tamper-proof event sourcing: Methods for accessing and validating previous events

### ISSUE THREAD

**Issue by gmkumar2005** [disintegrate-es/disintegrate#132]
How can I implement a tamper-proof event store using hash chaining? The idea is to compute and store the hash of an event (or a specific attribute) combined with the hash of the previous event. The tamper detection process works as follows:

    Retrieve the current and previous events.
    Concatenate the content of the current event with the digest from the previous event.
    Calculate the digest for this concatenated value.
    Compare the computed digest with the stored digest of the current event.
    If the digests match, proceed; otherwise, raise a tamper error.

In this use case, I want to specifically ensure the integrity of the `json_schema_string` attribute. I’m considering using the Decision::process trait as a starting point since it has access to the current command and state, which allows me to compute the digest for new events.

However, for verifying the digest, I need an API or mechanism to access the previous event. Could you provide guidance on implementing this approach?

```Rust
#[derive(Default, StateQuery, Clone, Debug, Serialize, Deserialize)]
#[state_query(DomainEvent)]
pub struct DefState {
    #[id]
    def_id: DefId,
    record_status: RecordStatus,
    json_schema_string: String,
    created_at: DateTime<Utc>,
    created_by: String,
    title: String,
    definitions: Vec<String>,
}
```

**Comment by luca-iachini** [comment]
Hi @gmkumar2005,
Thanks for trying out the library and sharing your use case!

One approach is to track the hash of the previous event in the `StateQuery`. Since `StateMutate` processes one event at a time, you can store the hash of the previous event in the `StateQuery` of the `Decision`. In the `mutate` implementation of the `StateQuery`, you can compute the hash of the current event, concatenate it with the hash of the previous event, and compare it to the hash of the event. If the hash is invalid, you could set a boolean flag like `is_tampered` in the `StateQuery`, which can then be checked in the `Decision::process` to determine if the events have been tampered. You can use the `prev_event_hash` of the `StateQuery` to emit new events with the proper hash.

```rust
#[derive(Default, StateQuery, Clone, Debug, Serialize, Deserialize)]
#[state_query(DomainEvent)]
pub struct DefState {
    #[id]
    def_id: DefId,
    record_status: RecordStatus,
    json_schema_string: String,
    created_at: DateTime<Utc>,
    created_by: String,
    title: String,
    definitions: Vec<String>,
    prev_event_hash: Hash,
    is_tampered: bool,
}

impl StateMutate for DefState {
    fn mutate(&mut self, event: Self::Event) {
        let event_hash = hash(event);
        self.prev_event_hash = hash(self.prev_event_hash + event_hash);
        if self.prev_event_hash != event.hash {
            self.is_tampered = true;
            // here you can track the tampered event for reporting it.
        }
    }
}
```

Let me know if that works for you!

**Comment by gmkumar2005** [comment]
Thanks for quick response.
This works because it will detect tamper across the entire history of the state. However it may not be suitable in cases where history is very large. For now I will go ahead with this solution.
Does this work as is in-case of a snapshot?

**Comment by luca-iachini** [comment]
Not quite, because snapshots require additional attention in your use case. You must ensure they have not been tampered with. A snapshot is created after every `x` events, with all subsequent events applied on top of it. You should verify both the integrity of the snapshot and that its `version` in the `snapshot` table has not been altered. (The snapshot version corresponds to the `event_id` of the last event included in the snapshot, which is used to retrieve new events occurring after it.)

To ensure integrity, you can implement a custom [StateSnapshotter](https://github.com/disintegrate-es/disintegrate/blob/6d1daaf8b7329d68dad4fe53d16550f3609b25d9/disintegrate/src/state_store.rs#L71C11-L71C27) inspired by the [PgSnapshotter](https://github.com/disintegrate-es/disintegrate/blob/6d1daaf8b7329d68dad4fe53d16550f3609b25d9/disintegrate-postgres/src/snapshotter.rs#L25), which includes your validation logic. For example, you could add an `HMAC` column and use it to validate both the snapshot payload and its version.

By the way, this approach could also be used for a custom `PgEventStore` that checks the `HMAC` of events in the same way.

This is how we would implement it, but you should think it through thoroughly to ensure that it covers all the guarantees you want to achieve.
